From a1f94e9856d3d24ddb5cb70d158c1f7a15135013 Mon Sep 17 00:00:00 2001
From: Martin Peres <martin.peres@linux.intel.com>
Date: Tue, 12 Jun 2018 12:12:27 +0300
Subject: [PATCH 06/10] modesetting: handle scanout damages at the crtc's
 refresh rate

This is not giving us tear-free, but we are getting close to it!

Signed-off-by: Martin Peres <martin.peres@linux.intel.com>
---
 hw/xfree86/drivers/modesetting/driver.c       | 17 +++--
 .../drivers/modesetting/drmmode_display.c     | 64 +++++++++++++++++++
 .../drivers/modesetting/drmmode_display.h     |  3 +
 3 files changed, 75 insertions(+), 9 deletions(-)

diff --git a/hw/xfree86/drivers/modesetting/driver.c b/hw/xfree86/drivers/modesetting/driver.c
index f6e1e59b3..dee6e5009 100644
--- a/hw/xfree86/drivers/modesetting/driver.c
+++ b/hw/xfree86/drivers/modesetting/driver.c
@@ -712,7 +712,7 @@ ms_dirty_get_ent(ScreenPtr screen, PixmapPtr slave_dst)
 }
 
 static void
-ms_refresh_all_scanout_buffers(ScreenPtr pScreen)
+ms_schedule_all_scanout_buffers_updates(ScreenPtr pScreen)
 {
     ScrnInfoPtr scrn = xf86ScreenToScrn(pScreen);
     xf86CrtcConfigPtr xf86_config = XF86_CRTC_CONFIG_PTR(scrn);
@@ -726,13 +726,13 @@ ms_refresh_all_scanout_buffers(ScreenPtr pScreen)
         if (!crtc->active || !drmmode_crtc->shadow_nonrotated)
             continue;
 
-        drmmode_update_scanout_buffer(crtc, drmmode_crtc->shadow_nonrotated);
+        /* schedule an async update, or revert to a synchronous copy on failure */
+        if (!drmmode_scanout_buffer_update_schedule(crtc,
+                                                    drmmode_crtc->shadow_nonrotated)) {
+            drmmode_update_scanout_buffer(crtc,
+                                          drmmode_crtc->shadow_nonrotated);
+        }
     }
-
-    /* force the buffer to be rendered to as quickly as possible */
-    #ifdef GLAMOR_HAS_GBM
-    glamor_finish(pScreen);
-    #endif
 }
 
 static void
@@ -751,8 +751,7 @@ msBlockHandler(ScreenPtr pScreen, void *timeout)
 
     ms_dirty_update(pScreen, timeout);
 
-    /* TODO: limit the blits to the CRTC's refresh rate to save memory BW */
-    ms_refresh_all_scanout_buffers(pScreen);
+    ms_schedule_all_scanout_buffers_updates(pScreen);
 }
 
 static void
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index 97f335e65..f368e04e0 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -1357,8 +1357,67 @@ drmmode_update_scanout_buffer(xf86CrtcPtr crtc, drmmode_shadow_scanout_ptr scano
 
     /* We have blitted all the damages, reset them */
     RegionEmpty(&scanout->screen_damage);
+
+    /* force the buffer to be rendered to as quickly as possible */
+    #ifdef GLAMOR_HAS_GBM
+    glamor_finish(pScreen);
+    #endif
+}
+
+struct scanout_buffer_update_args {
+    xf86CrtcPtr crtc;
+    drmmode_shadow_scanout_ptr scanout;
+};
+
+static void
+drmmode_scanout_buffer_update_handler(uint64_t frame, uint64_t usec,
+                                      void *data)
+{
+    struct scanout_buffer_update_args *args = data;
+
+    if (args->scanout) {
+        drmmode_update_scanout_buffer(args->crtc, args->scanout);
+    }
+    args->scanout->update_seq = 0;
+    free(args);
+}
+
+static void
+drmmode_scanout_buffer_update_abort(void *data)
+{
+    struct scanout_buffer_update_args *args = data;
+    args->scanout->update_seq = 0;
+    free(args);
+}
+
+Bool
+drmmode_scanout_buffer_update_schedule(xf86CrtcPtr crtc,
+                                       drmmode_shadow_scanout_ptr scanout)
+{
+    struct scanout_buffer_update_args *event_args;
+    int flip_seq = 0, ret;
+
+    if (scanout->update_seq)
+        return TRUE;
+
+    event_args = calloc(1, sizeof(*event_args));
+    if (!event_args)
+        return FALSE;
+
+    event_args->crtc = crtc;
+    event_args->scanout = scanout;
+
+    flip_seq = ms_drm_queue_alloc(crtc, event_args,
+                                  drmmode_scanout_buffer_update_handler,
+                                  drmmode_scanout_buffer_update_abort);
+
+    ret = ms_queue_vblank(crtc, MS_QUEUE_RELATIVE, 1, NULL, flip_seq);
+    scanout->update_seq = ret ? flip_seq : 0;
+
+    return ret;
 }
 
+
 static void
 drmmode_ConvertFromKMode(ScrnInfoPtr scrn,
                          drmModeModeInfo * kmode, DisplayModePtr mode)
@@ -2113,6 +2172,11 @@ drmmode_scanout_destroy(xf86CrtcPtr crtc, drmmode_shadow_scanout_ptr scanout)
     if (!scanout)
         return;
 
+    if (scanout->update_seq) {
+        ms_drm_abort_seq(crtc->scrn, scanout->update_seq);
+        scanout->update_seq = 0;
+    }
+
     if (scanout->pixmap) {
         scanout->pixmap->drawable.pScreen->DestroyPixmap(scanout->pixmap);
         scanout->pixmap = NULL;
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.h b/hw/xfree86/drivers/modesetting/drmmode_display.h
index aa0845a30..73ceced16 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.h
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.h
@@ -158,6 +158,7 @@ typedef struct {
     unsigned fb_id;
     PixmapPtr pixmap;
     RegionRec screen_damage;
+    uint32_t update_seq;
 } drmmode_shadow_scanout_rec, *drmmode_shadow_scanout_ptr;
 
 typedef struct {
@@ -283,6 +284,8 @@ Bool drmmode_SharedPixmapFlip(PixmapPtr frontTarget, xf86CrtcPtr crtc,
 void drmmode_DisableSharedPixmapFlipping(xf86CrtcPtr crtc, drmmode_ptr drmmode);
 
 void drmmode_update_scanout_buffer(xf86CrtcPtr crtc, drmmode_shadow_scanout_ptr scanout);
+Bool drmmode_scanout_buffer_update_schedule(xf86CrtcPtr crtc,
+                                            drmmode_shadow_scanout_ptr scanout);
 
 extern Bool drmmode_pre_init(ScrnInfoPtr pScrn, drmmode_ptr drmmode, int cpp);
 extern Bool drmmode_init(ScrnInfoPtr pScrn, drmmode_ptr drmmode);
-- 
2.25.1

